File: Java/BOJ/24511. queuestack.java
import java.io.*;
import java.util.*;

public class Main {    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        Deque <Integer> queue = new ArrayDeque<>();
        StringTokenizer st = new StringTokenizer(br.readLine()," ");
        int[] queuestack = new int[N];
        for (int i = 0; i < N; i ++) {
        	queuestack[i] = Integer.parseInt(st.nextToken());
        }
        int[] num = new int[N];
        st = new StringTokenizer(br.readLine(), " ");
        for(int i = 0 ; i < N; i++) {
        	num[i] = Integer.parseInt(st.nextToken());
        }
        int M = Integer.parseInt(br.readLine());
        int[] insert = new int[M];
        st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0 ; i < M; i++) {
        	insert[i] = Integer.parseInt(st.nextToken());
        }
        for(int i = 0 ; i< N; i ++) {
        	if (queuestack[i] == 0) {
        		queue.addFirst(num[i]);
        	}
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < M; i++){
        	queue.addLast(insert[i]);
        	sb.append(queue.pollFirst()).append(' ');
        }
        System.out.println(sb);
    }
}

File: Python/BOJ/24511 queuestack.py
import sys
from collections import deque

N = int(sys.stdin.readline())
queuestack = list(map(int, sys.stdin.readline().split()))
num = list(map(int, sys.stdin.readline().split()))
M = int(sys.stdin.readline())
insert = list(map(int, sys.stdin.readline().split()))
queue = deque()

for i in range(N):
    if queuestack[i] == 0:
        queue.appendleft(num[i])

result = []
for i in range(M):
    queue.append(insert[i])
    result.append(queue.popleft())
print(*result)
